[MRG+1] New text for the ValueError that is thrown by _check_param_grid method

b0noI opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue

<!-- Example: Fixes #1234 -->

N/A
#### What does this implement/fix? Explain your changes.

ValueError that is thrown by _check_param_grid method now includes parameters name when values of a parameter have incorrect type.

Old ValueError:
ValueError: Parameter values should be a list.

New ValueError:
ValueError: Parameter(key1) values should be a non-empty list.
#### Any other comments?

p.items() - creates a new list that includes tuples from the original dict on the Python2.x, however amount of a params should not be big, so it should not be a problem. On a Python 3.x items() crates iterator.

2 tests are broken:
sklearn.feature_extraction.tests.test_image.test_connect_regions ... FAIL
sklearn.feature_extraction.tests.test_image.test_connect_regions_with_grid ... FAIL
they have been broken before my commit. No new breakage is introduced by the commit itself.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

b0noI committed:
ValueError that is thrown by _check_param_grid method now includes parameters name when values of a parameter have incorrect type.

Old ValueError:
ValueError: Parameter values should be a list.

New ValueError:
ValueError: Parameter(key1) values should be a non-empty list.

b0noI committed:
ValueError that is thrown by _check_param_grid method now includes parameters name when values of a parameter have incorrect type.

Old ValueError:
ValueError: Parameter values should be a list.

New ValueError:
ValueError: Parameter(key1) values should be a non-empty list.

jnothman commented:
I don't mind something like this, but `sklean.grid_search

b0noI committed:
Merge branch 'master' of https://github.com/b0noI/scikit-learn

b0noI commented:
jnothman@ done, both places have been updated. Thank you for the feedback, I didn't know about deprecation.

PS: As soon as this one is merged, I will prepare PR that fixes support of range() in the method, since in Python3.X range does not return a list anymore. As a result some of the code is not working correctly after switching from Python2.X to 3.X

jnothman commented:
> PS: As soon as this one is merged, I will prepare PR that fixes support of range() in the method, since in Python3.X range does not return a list anymore. As a result some of the code is not working correctly after switching from Python2.X to 3.X

I'm not sure what you're referring to... Can you link to the misuse of `range`??

jnothman commented on sklearn/grid_search.py:
```diff
@@ -326,17 +326,18 @@ def _check_param_grid(param_grid):
             if len(v) == 0:
-                raise ValueError("Parameter values should be a non-empty "
-                                 "list.")
+                raise ValueError("Parameter({0}) values should be a non-empty "
```
I'd rather "Parameter ({0}) value should be ..."

b0noI replied:
done

b0noI committed:
"values" has been changed ot the "value" in the ValueError message.

b0noI commented:
> I'm not sure what you're referring to... Can you link to the misuse of range??

Simplified version of the logic from _check_param_grid method:
`v = range(0, 100)

check = [isinstance(v, k) for k in (list, tuple)]
print(True in check)
print(type(v))`

results on Python 2.X:
True
<type 'list'>

results on Python 3.X:
False
<class 'range'>

So some of the code is not working after migration to Python 3 =(

jnothman commented:
Yes, that does appear to be a separate bug, and is obviously not tested. We could make use of `collections.Sequence`.

b0noI commented:
jnothman@ yep, I can implement a tests for the usage of collections.Sequence. Probably will start with filing a bug =)

PS: I updated the text of the Errors

amueller commented on sklearn/grid_search.py:
```diff
@@ -326,17 +326,18 @@ def _check_param_grid(param_grid):
             check = [isinstance(v, k) for k in (list, tuple, np.ndarray)]
             if True not in check:
-                raise ValueError("Parameter values should be a list.")
+                raise ValueError("Parameter ({0}) value should be a list."
```
I have a hard time parsing this sentence. Currently when searching over parameter alpha, it says something like 

> Parameter (alpha) value should be a list.

That seems not very intuitive. Each parameter value for alpha should be a float.
Maybe " Parameter values for parameter {} need to be a sequence". ?

b0noI replied:
amueller@: your text look better, indeed. Thank you. PS: pull-request has been updated.

b0noI committed:
Text of the ValueError in _check_param_grid has been changed.

Old value: Parameter ({0}) value should be a list.
New value: Parameter values for parameter ({}) need to be a sequence.

b0noI commented:
jnothman@ anything else you want me to change in the request?

jnothman commented:
Well, I know for one thing that `{}` isn't supported in Python 2.6, which we still support and test. This line is clearly not covered by our current tests. Could you add a test for it?

b0noI committed:
Tests that are checking that _grid_search method throws correct ValueError have been added.

b0noI committed:
Test method test_grid_search_incorrect_param_grid has been renamed to test_grid_search_param_grid_includes_sequence_of_a_zero_length.

b0noI commented:
jnothman@ thank you! I did't know about the {} and Python 2.6. Error text has been updated. Plus I have added two tests that are covering cases when ValueError are thrown.

jnothman commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -165,6 +165,24 @@ def test_grid_search():
+        ValueError,
+        "Parameter values for parameter (C) need to be a sequence.",
+        GridSearchCV,
+        None,
```
I think you should provide an estimator here. The test case is unnecessarily pathological.

b0noI replied:
done

jnothman commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -165,6 +165,24 @@ def test_grid_search():
+    assert_raise_message(
+        ValueError,
+        "Parameter values for parameter (C) need to be a sequence.",
+        GridSearchCV,
```
We would not usually space things out this much. Put `GridSearchCV` and its parameters on the same line, I think.

b0noI replied:
done

b0noI committed:
MockClassifier has been added to the _grid_search tests.

jnothman commented on sklearn/grid_search.py:
```diff
@@ -326,17 +326,18 @@ def _check_param_grid(param_grid):
             check = [isinstance(v, k) for k in (list, tuple, np.ndarray)]
             if True not in check:
-                raise ValueError("Parameter values should be a list.")
+                raise ValueError("Parameter values for parameter ({0}) need to "
```
PEP8 line length

jnothman replied:
These should be strictly <80 characters

b0noI replied:
done

jnothman commented on sklearn/grid_search.py:
```diff
@@ -326,17 +326,18 @@ def _check_param_grid(param_grid):
             if len(v) == 0:
-                raise ValueError("Parameter values should be a non-empty "
-                                 "list.")
+                raise ValueError("Parameter values for parameter ({0}) need to "
```
PEP8 line length

b0noI replied:
done

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -327,17 +327,18 @@ def _check_param_grid(param_grid):
             check = [isinstance(v, k) for k in (list, tuple, np.ndarray)]
             if True not in check:
-                raise ValueError("Parameter values should be a list.")
+                raise ValueError("Parameter values for parameter ({0}) need to "
```
PEP8 line length

b0noI replied:
done

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -327,17 +327,18 @@ def _check_param_grid(param_grid):
             if len(v) == 0:
-                raise ValueError("Parameter values should be a non-empty "
-                                 "list.")
+                raise ValueError("Parameter values for parameter ({0}) need to "
```
PEP8 line length

b0noI replied:
done

b0noI committed:
Lines over 81 chracter length got reformatted.

jnothman commented:
LGTM. We usually require a second reviewer before merging.

ogrisel commented:
LGTM as well. Merging, thanks a lot @b0noI!
